src/DIRAC/Resources/IdProvider/IAMIdProvider.py

```diff
@@ -16,5 +16,13 @@ def getGroupScopes(self, group):
         """
         idPScope = getGroupOption(group, "IdPRole")
         if not idPScope:
-            idPScope = f"wlcg.groups:/{getVOForGroup(group)}/{group.split('_')[1]}"
+            vo = getVOForGroup(group)
+            if not vo:
+                return []
+
+            groupElements = group.split("_")
+            if len(groupElements) < 2:
+                return []
+
+            idPScope = f"wlcg.groups:/{vo}/{groupElements[1]}"
```

Comment:
Again, nothing to do with your PR, but this `"wlcg.groups"` seems just an hardcoded value...?

Reply:
Yes indeed.
The `wlcg.groups` scope is actually bound to the WLCG token profile that IAM supports.
But while I was playing with the IAM configuration, I noticed that it was not the only supported profile.
See here: https://indigo-iam.github.io/v/v1.8.0/docs/reference/configuration/jwt-profiles/
So the name should actually be something like: `WLCGIdProvider` or `IAMWLCGIdProvider`.
I think for now it is okay but we should keep that information in mind.

I made a small list of questions like that while working on these tests.
I should create some issues related to them, or may be create a small presentation to suggest some changes in the code.

Reply:
Then maybe rename the class to something more meaningful?

Does it mean that other VOs might need to employ a different profile?

Reply:
Since the discovery of the endpoints is well defined and standardized (https://openid.net/specs/openid-connect-discovery-1_0.html), there should not be any problem to retrieve the endpoints to interact with different IdProviders (meaning we would use `OAuth2IdProvider` to interact with `Indigo IAM`, `CheckIn`, ...). The main difference is the token profile used.

`VO_a` could theoretically use `Indigo IAM` with the `WLCG` token profile, while `VO_b` could use `Indigo IAM` with the `AARC` token profile and `VO_c` would use `CheckIn` with the `AARC` token profile.
Thus, I propose the following structure:

```
- Resources
  - IdPClient (currently IdProvider)
    - OAuth2IdPClient (currently OAuth2IdProvider - would embed a TokenProfile object)
    - TokenProfile
      - TokenProfile (abstract class defining getGroupScope() and getScopeGroup())
      - WLCGTokenProfile (currently IAMIdProvider)
      - AARCTokenProfile (currently CheckInProvider)
```

What do you think?
@atsareg

Reply:
I agree in general with the following remarks:
1. I do not think that IdProvider should change to IdPClient. All the DIRAC resources are actually clients of a particular third party service, however we name corresponding classes by rather the service nature related names. So, let's be consistent.
2. TokenProfile should be then made as loadable through ObjectLoader to be possible to develop these objects in extensions.

Reply:
1. I agree that most DIRAC resources are actually clients of a particular third party service. But I think this case is a bit different:
- Generally `IdProvider` refers to an `IAM` or `CheckIn` instance. But multiple OAuth2 clients can be defined and used to interact with a same "IdProvider" (`IAM` or `CheckIn`).
- `Client` is actually a standard OAuth2 term.

2. Yes sure.

Reply:
1. I think as with other resources we need here hierarchical description with corresponding objects in the code. Then the configuration should have rather the following structure:

    Resources
    {
        IdProviders
        {
            CheckIn
            {
                issuer = ...
                ...
                Clients
                {
                    Biomed
                    {
                        ClientID = ...
                        ClientSecret = ...
                        TokenHandler = BiomedTokenHandler
                        VO = biomed 
                    }
                    Galaxy
                    {
                        ClientID = ...
                        ClientSecret = ...
                        TokenHandler = DteamTokenHandler
                        VO = usegalaxy 
                    }
                } 
            } 
            WLCG_IAM
            {
                issuer = ...
                Clients
                {
                    DTeam
                    {
                        ...
                    }
                } 
            }
        }
    }

So, IdProvider stays and TokenProfile becomes IdPClient which has special methods for managing token profiles. The IdPClients can be configured to be used by limited number of VOs. The VO description in the Registry can have reference to the IdProvider as for example:

   Registry
   {
      VO
      {
         biomed
         {
            IdProvider = Check_In:::Biomed
         }
      }
   }

Note also that we should have a generic IdPClient class where group<->scopes mapping rules are simply defined in the configuration as one-to-one pairs. This should be better defined in the Registry, e.g.

    Registry
   {
      VO
      {
         biomed
         {
            IdProvider = Check_In
            TokenProfile
            {
                biomed_user = eduperson_entitlement?value=urn:mace:egi.eu:group:biomed:role=member#aai.egi.eu
                biomed_pilot = eduperson_entitlement?value=urn:mace:egi.eu:group:biomed:role=pilot#aai.egi.eu
            } 
         }
      }
   }

I think this will capture (almost) all the cases and no specialized TokenProfile classes will be needed at all. We can leave only with standard OAuth2IdProvider class configurable for all the VO needs.

Reply:
Well, the scope<->group definition is now in the Registry/Group/{group}/IdPRole . This was meant for defining extra scopes, e.g. compute scopes. But this can contain everything necessary for  scope<->group mapping. Should be better renamed to IdPMapping or IdPScope. Then the mapping will be done only in Group definition rather than in the VO section. BTW, this will allow diracAdmin group to survive.

Reply:
Example1: 

> I think as with other resources we need here hierarchical description with corresponding objects in the code. Then the configuration should have rather the following structure

Ok I agree with the structure, it seems better.
Just a few remarks to make sure we are on the same page (correct me if I don't understand correctly): 
- An `IdProviders` subsection (`CheckIn` or `WLCG_IAM` in your first example) corresponds to an instance of an IdP, which could handle, IIUC, one or many VOs.
- In your first example, you defined one client per VO (`Biomed`, `Galaxy`), but we could actually have multiple clients for a same VO: `Biomed_Pilot` the client used to submit pilots and `Biomed_Admin` the client used to perform token exchange for instance. 
- In your first example, the `TokenHandler` option corresponds to the token profile.

Example2:
> So, IdProvider stays and TokenProfile becomes IdPClient which has special methods for managing token profiles. The IdPClients can be configured to be used by limited number of VOs. The VO description in the Registry can have reference to the IdProvider as for example

Not sure to understand why `TokenProfile` becomes `IdPClient`? What is `IdPClient`? 

Does your second example mean that we have one client per VO? (+ another one defined in `Operations/<VO>/Pilot`?)
Because we are dealing with a OAuth2 client, may be the CS option should be named `IdProviderClient` instead of `IdProvider`, what do you think? 

Example3:

> Note also that we should have a generic IdPClient class where group<->scopes mapping rules are simply defined in the configuration as one-to-one pairs. This should be better defined in the Registry
> I think this will capture (almost) all the cases and no specialized TokenProfile classes will be needed at all. We can leave only with standard OAuth2IdProvider class configurable for all the VO needs.

Does it mean that we would have no `TokenProfile` class? Would the specificity of a token profile be entirely defined in the configuration then?
It seems like the `TokenProfile` section in your example correspond to group-based scopes. What if a VO does not support group-based scope? They would then define something like `biomed_pilot = compute.read compute.write...`. Therefore, I think that the section should actually be named `GroupScope`.

But the question is: isn't it the purpose of the `IdPRole` option (defined in `Groups/<VO>/IdPRole`)?

Except from that, if we have one token profile per VO, then may be it is better to define the token profile specific scopes in the configuration.
But I am not sure of that: for instance in EGI, VOs and `CheckIn` are using the `AARC` token profile, am I wrong?
LHC experiments use the `WLCG` token profile.
If multiple VOs use the same token profile, I think it is easier for them to add the name of a token profile rather than defining manually the token profile specific scopes in the CS (the `eduperson_entitlement` scope is rather complex to write).

Here is a suggestion combining my remarks:

```
Registry
{
  VO
  {
    biomed
    {
      IdProviderClient = Check_In:::Biomed_User
    }
    galaxy
    {
      IdProviderClient = Check_In:::Galaxy_User
    }  
  }
  Groups
  {
    biomed_user
    {
      # this would be parsed by a AARCTokeProfile instance
      GroupScope = biomed_member
      # if there is one token profile per VO, which does not seem to be the case, here is an alternative
      # IdPScope = eduperson_entitlement?value=urn:mace:egi.eu:group:biomed:role=member#aai.egi.eu
    }
    galaxy_admin
    {
      AttributeScope = compute.read compute.write ...
    }  
  }
}
Resources
 {
    IdProviders
    {
        CheckIn
        {
            issuer = ...
            ...
            Clients
            {
                Biomed_User
                {
                    ClientID = ...
                    ClientSecret = ...
                    TokenProfile = AARC
                    VO = biomed 
                }
                Biomed_Admin
               {
                   ClientID = ...
                   ClientSecret = ...
                   # the token profile can be defined at the level of the client, I don't think it would have any sense
                   # but it is technically possible
                   TokenProfile = WLCG
                   VO = biomed
               }
            } 
        } 
        WLCG_IAM
        {
            issuer = ...
            Clients
            {
                DTeam
                {
                    ...
                }
            } 
        }
    }
  }
```

Should this be included in this PR?
May be we should open an issue to further discuss about this?

Reply:
To summarize the outcome of our discussion with Alexandre on zoom, the following example of the configuration captures  the agreed schema:

```
Registry
{
  VO
  {
    biomed
    {
      IdProvider = CheckIn
      IdPClient = Biomed
    }
    dteam
    {
      IdProvider = WLCG_IAM
      IdPClient = Dteam
    }  
  }
  Groups
  {
    biomed_user
    {
      # The group uses Biomed IdPClient as defined in the biomed VO section 
      # this would be explicitly mapped by OAuth2IdProvider.getUserGroups(in parseIdPMapping),
      # no coded rules or regexps are used
      IdPMapping = eduperson_entitlement:urn:mace:egi.eu:group:registry:biomed:role=member#aai.egi.eu
      # No special other scopes for this group (so far)
    }
    biomed_pilot
    {
       IdPMapping = eduperson_entitlement:urn:mace:egi.eu:group:registry:biomed:role=pilot#aai.egi.eu
       Scope = compute.read compute.write ...
       # By default groups are using the IdProvider defined for the VO. In some cases a different client of the same
       # identity provider can be used
       IdPClient = BiomedPilot 
    }  
    dteam_user
    {
       IdPMapping = wlcg.groups:/dteam/user
    }
  }
}
Resources
 {
    IdProviders
    {
        CheckIn
        {
            issuer = ...
            Grant = device, refresh_token, authorization_code 
            ...
            Clients
            {
                Biomed
                {
                    ClientID = ...
                    ClientSecret = ...
                }
                BiomedPilot
               {
                   ClientID = ...
                   ClientSecret = ...
                   # overriding grant definition of the IdProvider
                   Grant = client_credentials
               }
            } 
        } 
        WLCG_IAM
        {
            issuer = ...
            Clients
            {
                DTeam
                {
                    ...
                }
            } 
        }
    }
  }
```

There are a couple of additions: 1. possibility to define IdProvider Client per group, group can not define another IdProvider than its VO; 2. Grant type that a given Client uses - this is to specify that pilot tokens are obtained with client_crdentials, for example.

Reply:
2. So IIUC, if one grant is defined, then we use it to get a token, right? If multiple ones are defined, what do we do? Is it only used to check that a requested grant type is allowed?

There is one missing point in your example: how to extract the group name from the right claim (`WLCG`: `wlcg.groups`).
We conclude that we should have a `Claim` or `GroupClaim` option defined in an `IdProvider` section (which could be overridden in a `Client` subsection)

Here is an additional example to take into account: 
I read the Indigo IAM documentation about token profile again, and it seems that a same OAuth2 client might distribute different tokens according to the group-based scope provided.
What if multiple small VOs use the same client but wants to use different token profiles?
Should the `GroupClaim` option be defined in `Resources/IdProviders/<IdProvider>/Clients/<Client>` or in `Registry/VO/<VO>`?

Reply:
For the grant preference I think we use client_credentials if it is specified in the configuration of the client, otherwise "normal" flows as it is now. 
As for the claim resolution - isn't it enough to take the claim name as the part of IdPMapping before the first colon ?

Reply:
In case of CheckIn the group claim is in the scope actually as for example eduperson_entitlement:urn:mace:egi.eu:group:training.egi.eu:role=member#aai.egi.eu

I guess WLCG IAM does the same ? So then we can look for the IdPMapping in the scopes simply without even extracting the group claim name

Reply:
> For the grant preference I think we use client_credentials if it is specified in the configuration of the client, otherwise "normal" flows as it is now.

So this could be a way to know whether a certain VO would like to use the `client_credentials` flow to submit pilots for instance, couldn't it?
But is there any advantage to set the other flows in this `Grant` option?
One managing the OAuth2 client has to think about updating the DIRAC configuration if there are changes.

> As for the claim resolution - isn't it enough to take the claim name as the part of IdPMapping before the first colon ? Then this answers also the last question

I was thinking about that yes. For the moment, I think it should be enough.